README
NAME
    Map::Tube::Graph - Graph output for Map::Tube.

SYNOPSIS
     use Map::Tube::Graph;
     my $obj = Map::Tube::GraphViz->new(%params);
     my $graph = $obj->graph;

METHODS
    "new(%params)"
             Constructor.

            *       "callback_edge"

                     Edge callback.
                     Default value is this:
                     sub { 
                             my ($self, $node, $link) = @_;
                             $self->{'graph'}->add_edge($node->id, $link);
                             return;
                     }

            *       "callback_vertex"

                     Vertex callback.
                     Default value is this:
                     sub { 
                             my ($self, $node) = @_;
                             $self->{'graph'}->add_vertex($node->id);
                             return;
                     }

            *       "graph"

                     Graph object.
                     Default value is Graph->new.

            *       "tube"

                     Map::Tube object.
                     It is required.
                     Default value is undef.

    "graph()"
             Get Graph object.
             Returns Graph object.

ERRORS
     new():
             Cannot load UNIVERSAL::DOES module.
             Parameter 'tube' is required.
             Parameter 'tube' must be 'Map::Tube' object.
             From Class::Utils::set_params():
                     Unknown parameter '%s'.

EXAMPLE
     use strict;
     use warnings;

     use English;
     use Error::Pure qw(err);
     use Map::Tube::Graph;

     # Arguments.
     if (@ARGV < 1) {
             print STDERR "Usage: $0 metro\n";
             exit 1;
     }
     my $metro = $ARGV[0];
 
     # Object.
     my $class = 'Map::Tube::'.$metro;
     eval "require $class;";
     if ($EVAL_ERROR) {
             err "Cannot load '$class' class.",
                     'Error', $EVAL_ERROR;
     }
 
     # Metro object.
     my $tube = eval "$class->new";
     if ($EVAL_ERROR) {
             err "Cannot create object for '$class' class.",
                     'Error', $EVAL_ERROR;
     }
 
     # GraphViz object.
     my $g = Map::Tube::Graph->new(
             'tube' => $tube,
     );
 
     # Get graph.
     my $graph = $g->graph;

     # Output.
     print $graph."\n";

     # Output without arguments like:
     # Usage: /tmp/SZXfa2g154 metro

     # Output with 'Prague' argument like:
     # A02-A01,A02-A03,A04-A03,A04-MUSTEK,A07-A08,A07-MUZEUM,A08-A09,A09-A10,A11-A10,A11-A12,A13-A12,B01-B02,B03-B02,B03-B04,B05-B04,B05-B06,B07-B06,B08-B07,B08-B09,B10-B09,B10-B11,B11-B12,B17-B16,B17-B18,B18-B19,B19-B20,B21-B20,B21-B22,B22-B23,B24-B23,C01-C02,C03-C02,C04-C03,C04-C05,C06-C05,C06-C07,C09-MUZEUM,C11-C12,C11-MUZEUM,C13-C12,C13-C14,C14-C15,C16-C15,C16-C17,C18-C17,C18-C19,C20-C19,FLORENC-B14,FLORENC-B16,FLORENC-C07,FLORENC-C09,MUSTEK-B12,MUSTEK-B14,MUSTEK-MUZEUM

DEPENDENCIES
    Class::Utils, English, Error::Pure, Graph, List::Util, Scalar::Util.

    UNIVERSAL::DOES if doesn't exists in Perl.

SEE ALSO
    Map::Tube
        Core library as Role (Moo) to process map data.

    Map::Tube::GraphViz
        GraphViz output for Map::Tube.

    Map::Tube::Plugin::Graph
        Graph plugin for Map::Tube.

    Map::Tube::Text::Table
        Table output for Map::Tube.

    Task::Map::Tube
        Install the Map::Tube modules.

REPOSITORY
    <https://github.com/michal-josef-spacek/Map-Tube-Graph>

AUTHOR
    Michal Josef Špaček <mailto:skim@cpan.org>

    <http://skim.cz>

LICENSE AND COPYRIGHT
    © 2014-2023 Michal Josef Špaček

    Artistic License

    BSD 2-Clause License

VERSION
    0.07



Changes
0.07

0.06 2023-11-30T19:05:50+01:00
 - Fix Makefile.PL for new version of Perl and no default »use lib '.'«.
 - Improve LICENSE AND COPYRIGHT section in doc.
 - Move bugtracker to github.
 - Remove simple comments.
 - Rewrite from List::MoreUtils to List::Util.
 - Update Module::Install to 1.21 version.
 - Update author github username.
 - Update author name.
 - Update copyright years.

0.05 2015-09-29T10:29:40+02:00
 - Add Map::Tube::Plugin::Graph to SEE ALSO section in doc.
 - Improve SEE ALSO section in doc.
 - Move author tests to xt/ directory.
 - Update Graph dependency to 0.9702 version.
 - Update Module::Install to 1.16 version.

0.04 2015-01-17T10:37:20+01:00
 - Fix usage of UNIVERSAL::DOES() in Perl older than 5.10.1.

0.03 2015-01-15T07:41:41+01:00
 - Add tests for constructor.
 - Fix does() to UNIVERSAL::DOES().
 - Fix test dependencies.

0.02 2015-01-11T12:03:46+01:00
 - Add tests for constructor.
 - Fix dependencies.
 - Remove obsolete test requirement.

0.01 2015-01-11T11:42:28+01:00
 - First version


Makefile.PL
use lib '.';
use strict;
use warnings;

use English;
use inc::Module::Install;

# Definition.
abstract 'Graph interface to the Map::Tube.';
author 'Michal Josef Spacek <skim@cpan.org>';
author_requires 'English' => 0;
author_requires 'File::Object' => 0.08;
author_requires 'Test::More' => 0;
author_requires 'Test::NoWarnings' => 0;
author_requires 'Test::Pod' => 0;
author_requires 'Test::Pod::Coverage' => 0;
license 'bsd';
license 'artistic_2';
name 'Map-Tube-Graph';
readme_from 'Graph.pm';
recursive_author_tests('xt');
requires 'Class::Utils' => 0.06;
requires 'Error::Pure' => 0.22;
requires 'Graph' => 0.9702;
requires 'List::Util' => 1.33;
requires 'Scalar::Util' => 0;
if ($PERL_VERSION lt v5.10.1) {
	requires 'UNIVERSAL::DOES' => 0;
}
requires 'perl' => '5.6.0';
resources 'bugtracker' => 'https://github.com/michal-josef-spacek/Map-Tube-Graph/issues';
resources 'homepage' => 'https://github.com/michal-josef-spacek/Map-Tube-Graph';
resources 'repository' => 'git://github.com/michal-josef-spacek/Map-Tube-Graph';
test_requires 'English' => 0;
test_requires 'Error::Pure::Utils' => 0.22;
test_requires 'Graph::Undirected' => 0;
test_requires 'Map::Tube::Prague' => 0;
test_requires 'Test::MockObject' => 0;
test_requires 'Test::More' => 0;
test_requires 'Test::NoWarnings' => 0;
tests_recursive;
version '0.07';

# Run.
WriteAll();


Graph.pm
package Map::Tube::Graph;

use strict;
use warnings;

use Class::Utils qw(set_params);
use English;
use Error::Pure qw(err);
use Graph;
use List::Util 1.33 qw(none);
use Scalar::Util qw(blessed);

our $VERSION = 0.07;

# Constructor.
sub new {
	my ($class, @params) = @_;

	# Create object.
	my $self = bless {}, $class;

	# Edge callback.
	$self->{'callback_edge'} = sub {
		my ($self, $node, $link) = @_;
		$self->{'graph'}->add_edge($node->id, $link);
		return;
	};

	# Vertex callback.
	$self->{'callback_vertex'} = sub {
		my ($self, $node) = @_;
		$self->{'graph'}->add_vertex($node->id);
		return;
	};

	# Graph object.
	$self->{'graph'} = undef;

	# Map::Tube object.
	$self->{'tube'} = undef;

	# Process params.
	set_params($self, @params);

	# Check Map::Tube object.
	if (! defined $self->{'tube'}) {
		err "Parameter 'tube' is required.";
	}
	if (! defined &UNIVERSAL::DOES) {
		eval {
			require UNIVERSAL::DOES;
		};
		if ($EVAL_ERROR) {
			err 'Cannot load UNIVERSAL::DOES module.';
		}
	}
	if (! blessed($self->{'tube'})
		|| ! $self->{'tube'}->DOES('Map::Tube')) {

		err "Parameter 'tube' must be 'Map::Tube' object.";
	}

	# Graph object.
	if (! defined $self->{'graph'}) {
		$self->{'graph'} = Graph->new;
	}

	# Object.
	return $self;
}

# Get graph.
sub graph {
	my $self = shift;
	foreach my $node (values %{$self->{'tube'}->nodes}) {
		$self->{'callback_vertex'}->($self, $node);
	}
	my @processed;
	foreach my $node (values %{$self->{'tube'}->nodes}) {
		foreach my $link (split m/,/ms, $node->link) {
			if (none {
				($_->[0] eq $node->id && $_->[1] eq $link) 
				|| 
				($_->[0] eq $link && $_->[1] eq $node->id)
				} @processed) {

				$self->{'callback_edge'}->($self, $node, $link);
				push @processed, [$node->id, $link];
			}
		}
	}
	return $self->{'graph'};
}

1;

__END__

=encoding utf8

=head1 NAME

Map::Tube::Graph - Graph output for Map::Tube.

=head1 SYNOPSIS

 use Map::Tube::Graph;
 my $obj = Map::Tube::GraphViz->new(%params);
 my $graph = $obj->graph;

=head1 METHODS

=over 8

=item C<new(%params)>

 Constructor.

=over 8

=item * C<callback_edge>

 Edge callback.
 Default value is this:
 sub { 
         my ($self, $node, $link) = @_;
         $self->{'graph'}->add_edge($node->id, $link);
         return;
 }

=item * C<callback_vertex>

 Vertex callback.
 Default value is this:
 sub { 
         my ($self, $node) = @_;
         $self->{'graph'}->add_vertex($node->id);
         return;
 }

=item * C<graph>

 Graph object.
 Default value is Graph->new.

=item * C<tube>

 Map::Tube object.
 It is required.
 Default value is undef.

=back

=item C<graph()>
 
 Get Graph object.
 Returns Graph object.

=back

=head1 ERRORS

 new():
         Cannot load UNIVERSAL::DOES module.
         Parameter 'tube' is required.
         Parameter 'tube' must be 'Map::Tube' object.
         From Class::Utils::set_params():
                 Unknown parameter '%s'.

=head1 EXAMPLE

 use strict;
 use warnings;

 use English;
 use Error::Pure qw(err);
 use Map::Tube::Graph;

 # Arguments.
 if (@ARGV < 1) {
         print STDERR "Usage: $0 metro\n";
         exit 1;
 }
 my $metro = $ARGV[0];
 
 # Object.
 my $class = 'Map::Tube::'.$metro;
 eval "require $class;";
 if ($EVAL_ERROR) {
         err "Cannot load '$class' class.",
                 'Error', $EVAL_ERROR;
 }
 
 # Metro object.
 my $tube = eval "$class->new";
 if ($EVAL_ERROR) {
         err "Cannot create object for '$class' class.",
                 'Error', $EVAL_ERROR;
 }
 
 # GraphViz object.
 my $g = Map::Tube::Graph->new(
         'tube' => $tube,
 );
 
 # Get graph.
 my $graph = $g->graph;

 # Output.
 print $graph."\n";

 # Output without arguments like:
 # Usage: /tmp/SZXfa2g154 metro

 # Output with 'Prague' argument like:
 # A02-A01,A02-A03,A04-A03,A04-MUSTEK,A07-A08,A07-MUZEUM,A08-A09,A09-A10,A11-A10,A11-A12,A13-A12,B01-B02,B03-B02,B03-B04,B05-B04,B05-B06,B07-B06,B08-B07,B08-B09,B10-B09,B10-B11,B11-B12,B17-B16,B17-B18,B18-B19,B19-B20,B21-B20,B21-B22,B22-B23,B24-B23,C01-C02,C03-C02,C04-C03,C04-C05,C06-C05,C06-C07,C09-MUZEUM,C11-C12,C11-MUZEUM,C13-C12,C13-C14,C14-C15,C16-C15,C16-C17,C18-C17,C18-C19,C20-C19,FLORENC-B14,FLORENC-B16,FLORENC-C07,FLORENC-C09,MUSTEK-B12,MUSTEK-B14,MUSTEK-MUZEUM

=head1 DEPENDENCIES

L<Class::Utils>,
L<English>,
L<Error::Pure>,
L<Graph>,
L<List::Util>,
L<Scalar::Util>.

L<UNIVERSAL::DOES> if doesn't exists in Perl.

=head1 SEE ALSO

=over

=item L<Map::Tube>

Core library as Role (Moo) to process map data.

=item L<Map::Tube::GraphViz>

GraphViz output for Map::Tube.

=item L<Map::Tube::Plugin::Graph>

Graph plugin for Map::Tube.

=item L<Map::Tube::Text::Table>

Table output for Map::Tube.

=item L<Task::Map::Tube>

Install the Map::Tube modules.

=back

=head1 REPOSITORY

L<https://github.com/michal-josef-spacek/Map-Tube-Graph>

=head1 AUTHOR

Michal Josef Špaček L<mailto:skim@cpan.org>

L<http://skim.cz>

=head1 LICENSE AND COPYRIGHT

© 2014-2023 Michal Josef Špaček

Artistic License

BSD 2-Clause License

=head1 VERSION

0.07

=cut
